upgrade/upgrades: TestJSONForwardingIndexes failed

upgrade/upgrades.TestJSONForwardingIndexes [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10974659?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10974659?buildTab=artifacts#/) on master @ [208ba72d61e334380be3f0bcf7ab1ecb3fae6ebd](https://github.com/cockroachdb/cockroach/commits/208ba72d61e334380be3f0bcf7ab1ecb3fae6ebd):


```
I230719 14:18:56.610677 14872 upgrade/upgrademanager/manager.go:657  [T1,n1,client=127.0.0.1:51566,hostssl,user=root,migration-mgr] 2938  executing operation bump-cluster-version=1000023.1--1
I230719 14:18:56.610924 14872 upgrade/upgradecluster/cluster.go:114  [T1,n1,client=127.0.0.1:51566,hostssl,user=root,migration-mgr] 2939  executing bump-cluster-version=1000023.1--1 on nodes n{1}
I230719 14:18:56.611166 36684 server/migration.go:150  [T1,n1,bump-cluster-version] 2940  active cluster version setting is now 1000023.1--1 (up from 1000022.2-102)
I230719 14:18:56.611425 14872 upgrade/upgrademanager/manager.go:657  [T1,n1,client=127.0.0.1:51566,hostssl,user=root,migration-mgr] 2941  executing operation validate-cluster-version=1000023.1
I230719 14:18:56.611656 14872 upgrade/upgradecluster/cluster.go:114  [T1,n1,client=127.0.0.1:51566,hostssl,user=root,migration-mgr] 2942  executing validate-cluster-version=1000023.1 on nodes n{1}
I230719 14:18:56.612041 14872 upgrade/upgrademanager/manager.go:657  [T1,n1,client=127.0.0.1:51566,hostssl,user=root,migration-mgr] 2943  executing operation bump-cluster-version=1000023.1
I230719 14:18:56.612271 14872 upgrade/upgradecluster/cluster.go:114  [T1,n1,client=127.0.0.1:51566,hostssl,user=root,migration-mgr] 2944  executing bump-cluster-version=1000023.1 on nodes n{1}
I230719 14:18:56.612479 36702 server/migration.go:150  [T1,n1,bump-cluster-version] 2945  active cluster version setting is now 1000023.1 (up from 1000023.1--1)
I230719 14:18:56.614762 14872 util/log/event_log.go:32  [T1,n1,client=127.0.0.1:51566,hostssl,user=root] 2946 ={"Timestamp":1689776325275326498,"EventType":"set_cluster_setting","Statement":"SET CLUSTER SETTING version = $1","Tag":"SET CLUSTER SETTING","User":"root","PlaceholderValues":["'1000023.1'"],"SettingName":"version","Value":"1000023.1"}
I230719 14:18:56.615162 14872 sql/schemachanger/scbuild/build.go:55  [T1,n1,client=127.0.0.1:51566,hostssl,user=root] 2947  building declarative schema change targets for CREATE INDEX
W230719 14:18:56.618501 14872 sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:393  [T1,n1,client=127.0.0.1:51566,hostssl,user=root] 2948  failed building declarative schema change targets for CREATE INDEX with error: unimplemented: column j is of type jsonb and thus is not indexable
W230719 14:18:56.618501 14872 sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:393  [T1,n1,client=127.0.0.1:51566,hostssl,user=root] 2948 +HINT: You have attempted to use a feature that is not yet implemented.
W230719 14:18:56.618501 14872 sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:393  [T1,n1,client=127.0.0.1:51566,hostssl,user=root] 2948 +See: https://go.crdb.dev/issue-v/35730/v23.2
I230719 14:18:56.618947 14872 sql/schemachanger/scbuild/build.go:55  [T1,n1,client=127.0.0.1:51566,hostssl,user=root] 2949  building declarative schema change targets for CREATE INDEX
W230719 14:18:56.623067 14872 sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:393  [T1,n1,client=127.0.0.1:51566,hostssl,user=root] 2950  failed building declarative schema change targets for CREATE INDEX with error: unimplemented: column crdb_internal_idx_expr is of type jsonb and thus is not indexable
W230719 14:18:56.623067 14872 sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:393  [T1,n1,client=127.0.0.1:51566,hostssl,user=root] 2950 +HINT: You have attempted to use a feature that is not yet implemented.
W230719 14:18:56.623067 14872 sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:393  [T1,n1,client=127.0.0.1:51566,hostssl,user=root] 2950 +See: https://go.crdb.dev/issue-v/35730/v23.2
I230719 14:18:56.627308 14923 9@sql/pgwire/conn.go:367  [T1,n1,client=127.0.0.1:51566,hostssl,user=root] 2951  could not parse simple query: CREATE TABLE test.tbl_json_secondary_unique(
I230719 14:18:56.627308 14923 9@sql/pgwire/conn.go:367  [T1,n1,client=127.0.0.1:51566,hostssl,user=root] 2951 +    j JSONB,
I230719 14:18:56.627308 14923 9@sql/pgwire/conn.go:367  [T1,n1,client=127.0.0.1:51566,hostssl,user=root] 2951 +    UNIQUE tbl_json_secondary_uidx (j)
I230719 14:18:56.627308 14923 9@sql/pgwire/conn.go:367  [T1,n1,client=127.0.0.1:51566,hostssl,user=root] 2951 +)
I230719 14:18:56.639007 14872 upgrade/upgrademanager/manager.go:397  [T1,n1,client=127.0.0.1:51566,hostssl,user=root,migration-mgr] 2952  migrating cluster from 1000023.1 to 1000023.1-14 (stepping through [1000023.1-2 1000023.1-4 1000023.1-6 1000023.1-8 1000023.1-10 1000023.1-12 1000023.1-14])
I230719 14:18:56.639094 14872 upgrade/upgrademanager/manager.go:657  [T1,n1,client=127.0.0.1:51566,hostssl,user=root,migration-mgr] 2953  executing operation validate-cluster-version=1000023.1-14
I230719 14:18:56.639373 14872 upgrade/upgradecluster/cluster.go:114  [T1,n1,client=127.0.0.1:51566,hostssl,user=root,migration-mgr] 2954  executing validate-cluster-version=1000023.1-14 on nodes n{1}
W230719 14:18:56.663013 14872 upgrade/upgrademanager/manager.go:370  [T1,n1,client=127.0.0.1:51566,hostssl,user=root,migration-mgr] 2955  error encountered during version upgrade: verifying precondition for version 1000023.1-2: query-invalid_objects: crdb-internal-jobs-table: system-jobs-scan: column "job_type" does not exist
    json_forward_indexes_test.go:147: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/upgrade/upgrades_test/pkg/upgrade/upgrades/json_forward_indexes_test.go:147
        	Error:      	Received unexpected error:
        	            	pq: verifying precondition for version 1000023.1-2: query-invalid_objects: crdb-internal-jobs-table: system-jobs-scan: column "job_type" does not exist
        	Test:       	TestJSONForwardingIndexes
I230719 14:18:56.663988 13364 testutils/testcluster/testcluster.go:148  [-] 2956  TestCluster quiescing nodes
I230719 14:18:56.664127 14098 sql/stats/automatic_stats.go:572  [T1,n1] 2957  quiescing auto stats refresher
I230719 14:18:56.664154 14099 sql/stats/automatic_stats.go:624  [T1,n1] 2958  quiescing stats garbage collector
W230719 14:18:56.664163 36417 kv/kvserver/intentresolver/intent_resolver.go:826  [-] 2959  failed to gc transaction record: could not GC completed transaction anchored at /Table/15/1/883772412356329473/0: node unavailable; try another peer
I230719 14:18:56.664272 13610 server/start_listen.go:103  [T1,n1] 2960  server shutting down: instructing cmux to stop accepting
W230719 14:18:56.664716 13867 sql/sqlliveness/slinstance/slinstance.go:334  [T1,n1] 2961  exiting heartbeat loop
W230719 14:18:56.664741 13867 sql/sqlliveness/slinstance/slinstance.go:321  [T1,n1] 2962  exiting heartbeat loop with error: node unavailable; try another peer
E230719 14:18:56.664759 13867 server/server_sql.go:515  [T1,n1] 2963  failed to run update of instance with new session ID: node unavailable; try another peer
I230719 14:18:56.664805 14319 jobs/registry.go:1606  [T1,n1] 2964  AUTO SPAN CONFIG RECONCILIATION job 883772374920265729: stepping through state succeeded
E230719 14:18:56.664820 27904 jobs/registry.go:1004  [T1,n1] 2965  error getting live session: node unavailable; try another peer
W230719 14:18:56.682596 31847 jobs/adopt.go:534  [T1,n1] 2966  could not clear job claim: clear-job-claim: node unavailable; try another peer
W230719 14:18:56.682721 26450 jobs/adopt.go:534  [T1,n1] 2967  could not clear job claim: clear-job-claim: failed to send RPC: sending to all replicas failed; last error: ba: Scan [/Table/15/1/102,/Table/15/1/103), [txn: 94b87cd5], [can-forward-ts] RPC error: node unavailable; try another peer
W230719 14:18:56.682830 26457 jobs/adopt.go:534  [T1,n1] 2968  could not clear job claim: clear-job-claim: node unavailable; try another peer
W230719 14:18:56.685108 35783 jobs/adopt.go:534  [T1,n1] 2969  could not clear job claim: clear-job-claim: node unavailable; try another peer
W230719 14:18:56.688088 30556 jobs/adopt.go:534  [T1,n1] 2970  could not clear job claim: clear-job-claim: node unavailable; try another peer
W230719 14:18:56.697942 29501 jobs/adopt.go:534  [T1,n1] 2971  could not clear job claim: clear-job-claim: node unavailable; try another peer
I230719 14:18:56.701753 13364 server/server_controller_orchestration.go:263  [T1,n1] 2972  server controller shutting down ungracefully
I230719 14:18:56.701799 13364 server/server_controller_orchestration.go:274  [T1,n1] 2973  waiting for tenant servers to report stopped
W230719 14:18:56.701894 13364 server/server_sql.go:1675  [T1,n1] 2974  server shutdown without a prior graceful drain
--- FAIL: TestJSONForwardingIndexes (11.85s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/release-eng
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestJSONForwardingIndexes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-29932

upgrade/upgrades.TestJSONForwardingIndexes [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10988760?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10988760?buildTab=artifacts#/) on master @ [3033423ccce8b4167dc7621d66d069ad003c1eb5](https://github.com/cockroachdb/cockroach/commits/3033423ccce8b4167dc7621d66d069ad003c1eb5):


```
I230720 06:19:07.532142 13759 upgrade/upgrademanager/manager.go:657  [T1,n1,client=127.0.0.1:36360,hostssl,user=root,migration-mgr] 2938  executing operation bump-cluster-version=1000022.2-102
I230720 06:19:07.532400 13759 upgrade/upgradecluster/cluster.go:114  [T1,n1,client=127.0.0.1:36360,hostssl,user=root,migration-mgr] 2939  executing bump-cluster-version=1000022.2-102 on nodes n{1}
I230720 06:19:07.532684 36693 server/migration.go:150  [T1,n1,bump-cluster-version] 2940  active cluster version setting is now 1000022.2-102 (up from 1000022.2-101(fence))
I230720 06:19:07.534980 13759 upgrade/upgrademanager/manager.go:517  [T1,n1,client=127.0.0.1:36360,hostssl,user=root,migration-mgr] 2941  stepping through 1000023.1
I230720 06:19:07.535026 13759 upgrade/upgrademanager/manager.go:657  [T1,n1,client=127.0.0.1:36360,hostssl,user=root,migration-mgr] 2942  executing operation bump-cluster-version=1000023.1--1
I230720 06:19:07.535264 13759 upgrade/upgradecluster/cluster.go:114  [T1,n1,client=127.0.0.1:36360,hostssl,user=root,migration-mgr] 2943  executing bump-cluster-version=1000023.1--1 on nodes n{1}
I230720 06:19:07.535477 36665 server/migration.go:150  [T1,n1,bump-cluster-version] 2944  active cluster version setting is now 1000023.1--1 (up from 1000022.2-102)
I230720 06:19:07.535756 13759 upgrade/upgrademanager/manager.go:657  [T1,n1,client=127.0.0.1:36360,hostssl,user=root,migration-mgr] 2945  executing operation validate-cluster-version=1000023.1
I230720 06:19:07.535988 13759 upgrade/upgradecluster/cluster.go:114  [T1,n1,client=127.0.0.1:36360,hostssl,user=root,migration-mgr] 2946  executing validate-cluster-version=1000023.1 on nodes n{1}
I230720 06:19:07.536333 13759 upgrade/upgrademanager/manager.go:657  [T1,n1,client=127.0.0.1:36360,hostssl,user=root,migration-mgr] 2947  executing operation bump-cluster-version=1000023.1
I230720 06:19:07.536555 13759 upgrade/upgradecluster/cluster.go:114  [T1,n1,client=127.0.0.1:36360,hostssl,user=root,migration-mgr] 2948  executing bump-cluster-version=1000023.1 on nodes n{1}
I230720 06:19:07.536756 36687 server/migration.go:150  [T1,n1,bump-cluster-version] 2949  active cluster version setting is now 1000023.1 (up from 1000023.1--1)
I230720 06:19:07.539003 13759 util/log/event_log.go:32  [T1,n1,client=127.0.0.1:36360,hostssl,user=root] 2950 ={"Timestamp":1689833936240054411,"EventType":"set_cluster_setting","Statement":"SET CLUSTER SETTING version = $1","Tag":"SET CLUSTER SETTING","User":"root","PlaceholderValues":["'1000023.1'"],"SettingName":"version","Value":"1000023.1"}
I230720 06:19:07.539429 13759 sql/schemachanger/scbuild/build.go:55  [T1,n1,client=127.0.0.1:36360,hostssl,user=root] 2951  building declarative schema change targets for CREATE INDEX
W230720 06:19:07.542661 13759 sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:393  [T1,n1,client=127.0.0.1:36360,hostssl,user=root] 2952  failed building declarative schema change targets for CREATE INDEX with error: unimplemented: column j is of type jsonb and thus is not indexable
W230720 06:19:07.542661 13759 sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:393  [T1,n1,client=127.0.0.1:36360,hostssl,user=root] 2952 +HINT: You have attempted to use a feature that is not yet implemented.
W230720 06:19:07.542661 13759 sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:393  [T1,n1,client=127.0.0.1:36360,hostssl,user=root] 2952 +See: https://go.crdb.dev/issue-v/35730/v23.2
I230720 06:19:07.543126 13759 sql/schemachanger/scbuild/build.go:55  [T1,n1,client=127.0.0.1:36360,hostssl,user=root] 2953  building declarative schema change targets for CREATE INDEX
W230720 06:19:07.546777 13759 sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:393  [T1,n1,client=127.0.0.1:36360,hostssl,user=root] 2954  failed building declarative schema change targets for CREATE INDEX with error: unimplemented: column crdb_internal_idx_expr is of type jsonb and thus is not indexable
W230720 06:19:07.546777 13759 sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:393  [T1,n1,client=127.0.0.1:36360,hostssl,user=root] 2954 +HINT: You have attempted to use a feature that is not yet implemented.
W230720 06:19:07.546777 13759 sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:393  [T1,n1,client=127.0.0.1:36360,hostssl,user=root] 2954 +See: https://go.crdb.dev/issue-v/35730/v23.2
I230720 06:19:07.550655 14814 9@sql/pgwire/conn.go:367  [T1,n1,client=127.0.0.1:36360,hostssl,user=root] 2955  could not parse simple query: CREATE TABLE test.tbl_json_secondary_unique(
I230720 06:19:07.550655 14814 9@sql/pgwire/conn.go:367  [T1,n1,client=127.0.0.1:36360,hostssl,user=root] 2955 +    j JSONB,
I230720 06:19:07.550655 14814 9@sql/pgwire/conn.go:367  [T1,n1,client=127.0.0.1:36360,hostssl,user=root] 2955 +    UNIQUE tbl_json_secondary_uidx (j)
I230720 06:19:07.550655 14814 9@sql/pgwire/conn.go:367  [T1,n1,client=127.0.0.1:36360,hostssl,user=root] 2955 +)
I230720 06:19:07.558815 13759 upgrade/upgrademanager/manager.go:397  [T1,n1,client=127.0.0.1:36360,hostssl,user=root,migration-mgr] 2956  migrating cluster from 1000023.1 to 1000023.1-14 (stepping through [1000023.1-2 1000023.1-4 1000023.1-6 1000023.1-8 1000023.1-10 1000023.1-12 1000023.1-14])
I230720 06:19:07.558875 13759 upgrade/upgrademanager/manager.go:657  [T1,n1,client=127.0.0.1:36360,hostssl,user=root,migration-mgr] 2957  executing operation validate-cluster-version=1000023.1-14
I230720 06:19:07.559124 13759 upgrade/upgradecluster/cluster.go:114  [T1,n1,client=127.0.0.1:36360,hostssl,user=root,migration-mgr] 2958  executing validate-cluster-version=1000023.1-14 on nodes n{1}
W230720 06:19:07.580113 13759 upgrade/upgrademanager/manager.go:370  [T1,n1,client=127.0.0.1:36360,hostssl,user=root,migration-mgr] 2959  error encountered during version upgrade: verifying precondition for version 1000023.1-2: query-invalid_objects: crdb-internal-jobs-table: system-jobs-scan: column "job_type" does not exist
    json_forward_indexes_test.go:147: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/upgrade/upgrades_test/pkg/upgrade/upgrades/json_forward_indexes_test.go:147
        	Error:      	Received unexpected error:
        	            	pq: verifying precondition for version 1000023.1-2: query-invalid_objects: crdb-internal-jobs-table: system-jobs-scan: column "job_type" does not exist
        	Test:       	TestJSONForwardingIndexes
I230720 06:19:07.584454 13193 testutils/testcluster/testcluster.go:148  [-] 2960  TestCluster quiescing nodes
I230720 06:19:07.584659 14006 sql/stats/automatic_stats.go:572  [T1,n1] 2961  quiescing auto stats refresher
W230720 06:19:07.584683 36641 kv/kvserver/intentresolver/intent_resolver.go:826  [-] 2962  failed to gc transaction record: could not GC completed transaction anchored at /Table/15/1/883961192206630913: node unavailable; try another peer
I230720 06:19:07.584715 14007 sql/stats/automatic_stats.go:624  [T1,n1] 2963  quiescing stats garbage collector
W230720 06:19:07.584754 13940 sql/sqlliveness/slinstance/slinstance.go:334  [T1,n1] 2964  exiting heartbeat loop
W230720 06:19:07.584819 13940 sql/sqlliveness/slinstance/slinstance.go:321  [T1,n1] 2966  exiting heartbeat loop with error: node unavailable; try another peer
I230720 06:19:07.584818 14283 jobs/registry.go:1606  [T1,n1] 2965  AUTO SPAN CONFIG RECONCILIATION job 883961154528051201: stepping through state succeeded
E230720 06:19:07.584851 13940 server/server_sql.go:516  [T1,n1] 2967  failed to run update of instance with new session ID: node unavailable; try another peer
E230720 06:19:07.584869 31838 jobs/registry.go:1004  [T1,n1] 2968  error getting live session: node unavailable; try another peer
I230720 06:19:07.584897 13242 server/start_listen.go:103  [T1,n1] 2969  server shutting down: instructing cmux to stop accepting
W230720 06:19:07.591942 29473 jobs/adopt.go:534  [T1,n1] 2970  could not clear job claim: clear-job-claim: failed to send RPC: sending to all replicas failed; last error: ba: Scan [/Table/15/1/883961178860486657,/Table/15/1/883961178860486658), [txn: 7b659f3b], [can-forward-ts] RPC error: node unavailable; try another peer
W230720 06:19:07.592505 35019 jobs/adopt.go:534  [T1,n1] 2971  could not clear job claim: clear-job-claim: node unavailable; try another peer
I230720 06:19:07.596860 13193 server/server_controller_orchestration.go:263  [T1,n1] 2972  server controller shutting down ungracefully
I230720 06:19:07.596938 13193 server/server_controller_orchestration.go:274  [T1,n1] 2973  waiting for tenant servers to report stopped
W230720 06:19:07.597023 13193 server/server_sql.go:1676  [T1,n1] 2974  server shutdown without a prior graceful drain
--- FAIL: TestJSONForwardingIndexes (11.77s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestJSONForwardingIndexes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/jobs/BUILD.bazel
+++ b/pkg/jobs/BUILD.bazel
@@ -49,6 +49,7 @@ go_library(
         "//pkg/sql/catalog/descpb",
         "//pkg/sql/catalog/descs",
         "//pkg/sql/isql",
+        "//pkg/sql/pgwire/pgcode",
         "//pkg/sql/pgwire/pgerror",
         "//pkg/sql/protoreflect",
         "//pkg/sql/sem/builtins",

--- a/pkg/jobs/utils.go
+++ b/pkg/jobs/utils.go
@@ -19,6 +19,8 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
 	"github.com/cockroachdb/cockroach/pkg/kv"
 	"github.com/cockroachdb/cockroach/pkg/sql/isql"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/errors"
 )
@@ -152,4 +154,34 @@ func JobExists(
 		return false, err
 	}
 	return row != nil, nil
+}
+
+// IsJobTypeColumnDoesNotExistError returns true if the error is of the form
+// `column "job_type" does not exist`.
+func isJobTypeColumnDoesNotExistError(err error) bool {
+	return pgerror.GetPGCode(err) == pgcode.UndefinedColumn &&
+		strings.Contains(err.Error(), "column \"job_type\" does not exist")
+}
+
+// MaybeGenerateForcedRetryableError returns a
+// TransactionRetryWithProtoRefreshError that will cause the txn to be retried
+// if the error is because of an undefined job_type column.
+//
+// In https://github.com/cockroachdb/cockroach/issues/106762 we noticed that if
+// a query is executed with an AS OF SYSTEM TIME clause that picks a transaction
+// timestamp before the job_type migration, then parts of the jobs
+// infrastructure will attempt to query the job_type column even though it
+// doesn't exist at the transaction's timestamp.
+//
+// As a short term fix, when we encounter an `UndefinedColumn` error we
+// generate a synthetic retryable error so that the txn is pushed to a
+// higher timestamp at which the upgrade will have completed and the
+// `job_type` column will be visible. The longer term fix is being tracked
+// in https://github.com/cockroachdb/cockroach/issues/106764.
+func MaybeGenerateForcedRetryableError(ctx context.Context, txn *kv.Txn, err error) error {
+	if err != nil && isJobTypeColumnDoesNotExistError(err) {
+		return txn.GenerateForcedRetryableError(ctx, "synthetic error "+
+			"to push timestamp to after the `job_type` upgrade has run")
+	}
+	return err
 }

--- a/pkg/sql/crdb_internal.go
+++ b/pkg/sql/crdb_internal.go
@@ -1069,7 +1069,7 @@ func populateSystemJobsTableRows(
 		params...,
 	)
 	if err != nil {
-		return matched, err
+		return matched, jobs.MaybeGenerateForcedRetryableError(ctx, p.Txn(), err)
 	}
 
 	cleanup := func(ctx context.Context) {
@@ -1082,7 +1082,7 @@ func populateSystemJobsTableRows(
 	for {
 		hasNext, err := it.Next(ctx)
 		if !hasNext || err != nil {
-			return matched, err
+			return matched, jobs.MaybeGenerateForcedRetryableError(ctx, p.Txn(), err)
 		}
 
 		currentRow := it.Cur()

--- a/pkg/upgrade/upgrades/json_forward_indexes_test.go
+++ b/pkg/upgrade/upgrades/json_forward_indexes_test.go
@@ -25,7 +25,6 @@ import (
 )
 
 func TestJSONForwardingIndexes(t *testing.T) {
-	skip.WithIssue(t, 107169, "flaky test")
 	var err error
 	skip.UnderStressRace(t)
 	defer leaktest.AfterTest(t)()

